### ISSUE TITLE
Java::JavaLang::ArrayIndexOutOfBoundsException on jruby-9.3.2.0

### ISSUE THREAD

**Issue by klobuczek** [jruby/jruby#6984]
When running specs with simplecov on jruby-9.3.2.0 I'm getting the below problem. This does not happen when specs are run without coverage. On jruby-9.2.20.1 it works with or without coverage.

I have created a very minimal crippled down rails app to reproduce the problem: https://github.com/klobuczek/bug-app
You can see the executed GH actions here: https://github.com/klobuczek/bug-app/actions/runs/1655924098

```
Failures:

  1) BugMailer bug
     Failure/Error: Unable to find org/jruby/RubyKernel.java to read failed line

     LoadError:
       load error: mail/parsers/address_lists_parser -- java.lang.ArrayIndexOutOfBoundsException: -32624
     # org/jruby/RubyKernel.java:1017:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `block in require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:299:in `load_dependency'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/parsers.rb:6:in `<main>'
     # org/jruby/RubyKernel.java:1017:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `block in require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:299:in `load_dependency'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/parsers/content_type_parser.rb:6:in `<main>'
     # org/jruby/RubyKernel.java:1017:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `block in require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:299:in `load_dependency'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/elements/content_type_element.rb:3:in `<main>'
     # org/jruby/RubyKernel.java:1017:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `block in require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:299:in `load_dependency'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/activesupport-6.1.4.4/lib/active_support/dependencies.rb:332:in `require'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/fields/content_type_field.rb:39:in `element'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/fields/content_type_field.rb:33:in `parse'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/fields/content_type_field.rb:25:in `initialize'
     # org/jruby/RubyClass.java:893:in `new'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/field.rb:277:in `new_field'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/field.rb:266:in `create_field'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/field.rb:187:in `field'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/field.rb:239:in `method_missing'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/header.rb:186:in `[]='
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/message.rb:1213:in `default'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/mail-2.7.1/lib/mail/message.rb:613:in `content_type'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/actionmailer-6.1.4.4/lib/action_mailer/test_case.rb:102:in `set_expected_mail'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-rails-5.0.2/lib/rspec/rails/adapters.rb:113:in `block in setup'
     # org/jruby/RubyBasicObject.java:2673:in `instance_exec'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:455:in `instance_exec'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:365:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:529:in `block in run_owned_hooks_for'
     # org/jruby/RubyArray.java:1865:in `each'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:528:in `run_owned_hooks_for'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:615:in `block in run_example_hooks_for'
     # org/jruby/RubyArray.java:1947:in `reverse_each'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:614:in `run_example_hooks_for'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:484:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:502:in `run_before_example'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:261:in `block in run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:508:in `block in with_around_and_singleton_context_hooks'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:465:in `block in with_around_example_hooks'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:486:in `block in run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:350:in `call'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-rails-5.0.2/lib/rspec/rails/adapters.rb:75:in `block in MinitestLifecycleAdapter'
     # org/jruby/RubyBasicObject.java:2673:in `instance_exec'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:455:in `instance_exec'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:390:in `execute_with'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:628:in `block in run_around_example_hooks_for'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:350:in `call'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/hooks.rb:486:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:465:in `with_around_example_hooks'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:508:in `with_around_and_singleton_context_hooks'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example.rb:259:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example_group.rb:644:in `block in run_examples'
     # org/jruby/RubyArray.java:2642:in `map'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example_group.rb:640:in `run_examples'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/example_group.rb:606:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:121:in `block in run_specs'
     # org/jruby/RubyArray.java:2642:in `map'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:121:in `block in run_specs'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/configuration.rb:2067:in `with_suite_hooks'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:116:in `block in run_specs'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/reporter.rb:74:in `report'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:115:in `run_specs'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:89:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:71:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/lib/rspec/core/runner.rb:45:in `invoke'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/rspec-core-3.10.1/exe/rspec:4:in `<main>'
     # org/jruby/RubyKernel.java:1052:in `load'
     # /home/runner/.rubies/jruby-9.3.2.0/bin/rspec:25:in `<main>'
     # org/jruby/RubyKernel.java:1052:in `load'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/cli/exec.rb:58:in `kernel_load'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/cli/exec.rb:23:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/cli.rb:484:in `exec'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/cli.rb:31:in `dispatch'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/cli.rb:25:in `start'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/exe/bundle:48:in `block in <main>'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
     # /home/runner/.rubies/jruby-9.3.2.0/lib/ruby/gems/shared/gems/bundler-2.3.4/exe/bundle:36:in `<main>'
     # org/jruby/RubyKernel.java:1052:in `load'
     # /home/runner/.rubies/jruby-9.3.2.0/bin/bundle:23:in `<main>'
     # 
     #   Showing full backtrace because every line was filtered out.
     #   See docs for RSpec::Configuration#backtrace_exclusion_patterns and
     #   RSpec::Configuration#backtrace_inclusion_patterns for more information.
     # ------------------
     # --- Caused by: ---
     # Java::JavaLang::ArrayIndexOutOfBoundsException:
     #   -32624
     #   org.jruby.parser.ParserConfiguration.coverLine(ParserConfiguration.java:193)
```

**Environment Information**
  
Provide at least:
* JRuby version 9.3.20
* Operating system and platform Mac OSX and Ubuntu

Other relevant info you may wish to add:
* Installed or activated gems
* Application/framework version rails 6.1.4.4
* Environment variables

**Expected Behavior**

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.
* Provide an executable Ruby script or a link to an example repository.

**Actual Behavior**

* Describe or show the actual behavior.
* Provide text or screen capture showing the behavior.

**Comment by ahorek** [comment]
it fails since https://github.com/jruby/jruby/commit/2cfb230e9488758fca5789f1a2943057bc4a2f9d

but the problem is that file https://github.com/mikel/mail/blob/2-7-stable/lib/mail/parsers/address_lists_parser.rb is too large and lines after INT_MAX just overflows. The current limit is 32767 lines.

```if (i < 0) return;``` here would prevent the overflow https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/parser/ParserConfiguration.java#L190 but it also breaks correct coverage for these big files.

eventually, storing lines as bigintegers is probably the best solution, but it would require significant changes.

**Comment by enebo** [comment]
@klobuczek Thanks for the reproduction.  I suppose we should put in @ahorek negative value check to make sure people do not see things when we overflow but we also will need to increase how many lines we can support. 2 more bits would get us up to 128k lines but then decrease columns 8000.  I am loathe to give full values for this since it substantially increases our memory usage.  Perhaps upping to a long amount of space for the two won't affect memory that much and it would put this to rest.

We are thinking about 9.3.3.0 early next week so this will be fixed for that.

**Comment by enebo** [comment]
Fixed in #6991.  As it turns out the variables being tweaked only exist while parsing.  So line and column both get 32 bits now. 32bits is how much space we store line in after parsing so that is a very hard limit but how many people have files 2147483647 lines long?
